PadicsLoco/PadicsLoco.lean

```diff
@@ -0,0 +1,119 @@
+/-
+Copyright (c) 2024 Jou Glasheen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jou Glasheen, Kevin Buzzard
+-/
+
+import Mathlib.NumberTheory.Padics.RingHoms
+import Mathlib.Topology.MetricSpace.Ultra.Basic
+
+/-!
+# Padics totally bounded, compact, locally compact
+
+In this file, we prove that `Z_[p]` is totally bounded and compact,
+and that `ℚ_[p]` is locally compact.
+
+## Main results
+
+- `PadicInt.totally_bounded` : The set of p-adic integers `ℤ_[p]` is totally bounded.
+- `PadicInt.is_compact` : The set of p-adic integers `ℤ_[p]` is a compact topological space.
+- `Padic.locally_compact_space` : The set of p-adic numbers ℚ_[p] is a locally compact topological
+                                  space.
+
+## Notation
+
+ - `p` : Is a natural prime.
+
+## References
+
+Gouvêa, F. Q. (2020) p-adic Numbers An Introduction. 3rd edition.
+  Cham, Springer International Publishing
+-/
+
+variable (p : ℕ) [Fact (Nat.Prime p)]
+
+namespace PadicInt
+
+/--The set of p-adic integers `ℤ_[p]` is totally bounded, where (p : ℕ) [Fact (Nat.Prime p)].-/
+theorem totally_bounded :
+    TotallyBounded (Set.univ : Set ℤ_[p]) := by
+  apply Metric.totallyBounded_iff.2
+  intros ε he
+  apply PadicInt.exists_pow_neg_lt p at he
+  rcases he with ⟨k , hk⟩
+  let t0 := Finset.range ((p ^ k) : ℕ)
+  let t : Set ℤ_[p] := ((↑) : ℕ → ℤ_[p])'' t0
+  use t
+  constructor
+  · exact Set.toFinite t
+  · intros z _
+    simp only [PadicInt, Set.mem_iUnion, Metric.mem_ball, exists_prop]
+    use ((↑) : ℕ → ℤ_[p]) (PadicInt.appr z k)
+    constructor
+    · refine (Set.mem_image Nat.cast ↑t0 ((z.appr k) : ℤ_[p])).mpr ?h.left.a
+      simp only [Finset.mem_coe, Nat.cast_inj]
+      refine exists_eq_right.mpr ?h.left.a.a
+      refine Finset.mem_range.mpr ?h.left.a.a.a
+      exact PadicInt.appr_lt z k
+    · have h1 : dist z ((z.appr k) : ℤ_[p]) ≤ (p ^ (- (k : ℤ ))) := by
+        have h3 : (z ) - ↑(z.appr k) ∈ Ideal.span {(p  : ℤ_[p]) ^ k} := PadicInt.appr_spec k z
+        have h4 : ‖(z) - ↑(z.appr k)‖  ≤ (p : ℝ) ^ (- k : ℤ) :=
+          (PadicInt.norm_le_pow_iff_mem_span_pow (z - ((z.appr k) : ℤ_[p])) k).mpr h3
+        exact h4
+      exact lt_of_le_of_lt h1 hk
+
+/--The set of p-adic integers `ℤ_[p]` is a compact topological space,
+where (p : ℕ) [Fact (Nat.Prime p)].-/
+  theorem is_compact : IsCompact (Set.univ : Set ℤ_[p]) := by
+    apply isCompact_iff_totallyBounded_isComplete.2
+    constructor
+    · exact totally_bounded p
+    · exact complete_univ
```

Comment:
I think you probably want `CompactSpace` (the whole space is compact), not `IsCompact Set.univ`, here. It should also be declared as a typeclass instance (so the typeclass resolution machinery can use it). Perhaps:
```suggestion
/-- The set of p-adic integers `ℤ_[p]` is a compact topological space. -/
instance compactSpace : CompactSpace ℤ_[p] := by
  rw [← isCompact_univ_iff, isCompact_iff_totallyBounded_isComplete]
  exact ⟨totallyBounded p, complete_univ⟩
```